alf/environments/alf_gym3_wrapper.py

```diff
@@ -0,0 +1,441 @@
+        ]
+
+        return ds.TimeStep(
+            step_type=torch.as_tensor(step_type),
+            reward=torch.as_tensor(reward),
+            discount=torch.as_tensor(discount),
+            observation=observation,
+            env_id=torch.arange(self.batch_size),
+            prev_action=torch.as_tensor(action),
+            env_info=nest.map_structure(
+                lambda *values: torch.as_tensor(values), *trimmed_info))
+
+    # Implement abstract _reset
+    def _reset(self) -> ds.TimeStep:
+        """Implement ``_reset()`` for compatibility
+
+        Note that by default Gym3 environmenst cannot be reset. However, if
+        ``support_force_reset`` is set to True, action -1 will be sent to all
+        the sub-environments to reset them if the underlying environment follows
+        the convention.
+
+        Otherwise the reset will be ignored.
+
+        """
+        if self._support_force_reset:
+            self._gym3_env.act(np.array([-1] * self.batch_size))
+            self._prev_first = [False] * self.batch_size
+        else:
+            logging.warning('reset() ignored by AlfGym3Wrapper')
+
+        reward, observation, _ = self._gym3_env.observe()
+
+        time_step = self._create_time_step(
+            reward,
+            observation,
+            step_type=[ds.StepType.FIRST] * self.batch_size,
+            # Faking actions
+            action=nest.map_structure(
+                lambda spec: spec.numpy_zeros(outer_dims=(self.batch_size, )),
+                self.action_spec()))
+
+        return time_step
+
+    # Implement abstract _step
+    def _step(self, action) -> ds.TimeStep:
+        """Implement ``_step()`` for compatibility
+
+        A very important note here is that special treatment is done at the end
+        of each episode. Unlike Gym environments, Gym3 environments do not
+        return ``done=True`` immediately. Instead, ``first=True`` is returned
+        upon the ``observe()`` of the NEXT ``act()``, which is actually the
+        FIRST FRAME of a new episode in Gym3 sense.
+
+        With this wrapper we CHANGED that definition, so that the frame with
+        ``first=True`` actually becomes the LAST FRAME of the previous episode.
+        Because the observation of the new episode can be DRAMATICALLY DIFFERENT
+        from the actual last frame of the previous episode, the previous
+        observation is returned in this case.
+
+        Therefore to summarize, effectively we will
+
+        1. Repeat the end-of-episode observation twice for each episode.
+
+        2. Throw away the first frame of each episode, and use the second frame
+           as if it is the first frame.
+
+        Gym3's official Gym wrapper DID THE SAME.
+
+        """
+        _, prev_observation, _ = self._gym3_env.observe()
+
+        np_action = nest.map_structure(lambda x: x.cpu().numpy(), action)
+        self._gym3_env.act(np_action)
+
+        reward, observation, first = self._gym3_env.observe()
```

Comment:
Not sure how high the cost is to call observe(). If takes much time, then probably we can avoid x2 times of calls by first calling observation() in reset() and then each time just call once in _step().

Reply:
The `observation` is stored in the buffer so accessing it in a read-only way should not incur extra cost. I was optimizing more towards "less stateful wrapper" here.